### ISSUE TITLE
Evaluation with MCTS not consistent

### ISSUE THREAD

**Issue by dalkio** [neuronest/custom-alphazero#14]
Right now, evaluations with or without MCTS are not consistent at all:

```python
from src.evaluation.evaluate import evaluate_two_models
from src.utils import best_saved_model, set_gpu_index, init_model

set_gpu_index("-1")

best_model = best_saved_model("2020-09-06-180343")
random_model = init_model()

score, solver_score = evaluate_two_models(best_model, random_model)
score
>>> 0.8156028368794326

score, solver_score = evaluate_two_models(best_model, random_model, evaluate_with_mcts=True)
score
>>> 0.4460431654676259
```

_(tested with run_id="2020-09-06-180343", evaluation_iteration="iteration_688")_

Check if this is a bug or a proof of a faulty model (#13).

**Comment by dalkio** [comment]
[results_2020-09-06-180343.zip](https://github.com/neuronest/custom-alphazero/files/5183076/results_2020-09-06-180343.zip)

**Comment by yohskua** [comment]
Je n'ai pas vue de bug particulier. 

- Sans mcts le trained modèle gagne la plupart du temps car il a quand même appris à reconnaître des situations où il sait quel coup jouer pour gagner et le random modèle joue au hasard donc même quand il peut gagner il a peu de chances de jouer le bon coup
- Avec mcts, les situations où le random modèle peut gagner et aurait fait n'importe quoi sans mcts ne sont pas choisies car la policy presque uniforme du random modèle fait que tous les coups sont testés et si un coup mène à la victoire ça se verra dans Q et finalement dans PN et le random modèle avec mcts jouera le coup gagnant. 
- Avec mcts, le trained model a des priors trop forts, il ne voit même plus que l'adversaire gagnera au coup d'après dans certaines situations car, dans l'arbre, il ne tente même pas ensuite le coup qui aurait fait gagner l'adversaire et lui aurait back up un reward de -1. Les V values sont également mauvaises (aux alentours de 0) pour des situations évidemment mauvaises après un coup (par exemple trois croix XXX alignées pour l'adversaire).

Le réseau est trop sûr de lui dans des situations où il se trompe + mauvaise value, empêchant le mcts de parser intelligemment les différentes possibilités + le random modèle avec mcts est lui bien meilleur à jouer les bons coups s'il est sur le point de gagner.

A faire peut être : 

- learning rate plus bas pour ralentir l'overfit et le désapprentissage de situations de base
- augmenter la taille de la queue pour ne pas perdre trop tôt les situations évidentes découvertes au début
- augmenter le nombre de simulations pour le forcer à trouver les coups obvious quand le prior est élevé et faux
- Dirichlet noise pour forcer à davantage d'exploration quand les priors sont élevés
